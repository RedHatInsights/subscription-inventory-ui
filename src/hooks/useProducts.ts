import { useQuery, QueryObserverResult } from 'react-query';
import { useToken } from '../utilities/platformServices';

type Product = {
  name: string;
  productLine: string;
  quantity: number;
  sku: string;
  serviceLevel: string;
  serviceType: string;
  capacity: Capacity;
  subscriptions?: Subscription[];
};

type Subscription = {
  contractNumber: string;
  endDate: string;
  number: string;
  quantity: string;
  startDate: string;
  status: string;
};

interface ProductApiData {
  body: Product[];
}

type Capacity = {
  name: string;
  quantity: string;
};

const UoMNameOrder = ['Cores', 'Nodes', 'Sockets'];

const fetchProductData =
  (jwtToken: Promise<string>) =>
  async (filter: string): Promise<Product[]> => {
    const response = await fetch(`/api/rhsm/v2/products?status=${filter}`, {
      headers: { Authorization: `Bearer ${await jwtToken}` }
    });

    const productResponseData: ProductApiData = await response.json();

    return productResponseData.body;
  };

const useProducts = (filter: string): QueryObserverResult<Product[], unknown> => {
  const jwtToken = useToken();
  return useQuery(`products.${filter}`, () => fetchProductData(jwtToken)(filter));
};

export { Product, Subscription, Capacity, UoMNameOrder, useProducts as default };
