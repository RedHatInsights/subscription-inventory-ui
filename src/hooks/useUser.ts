import { useQuery } from 'react-query';
import { useAuthenticateUser, useUserRbacPermissions } from '../utilities/platformServices';

interface User {
  accountNumber: string;
  canReadProducts: boolean;
  isOrgAdmin: boolean;
}

const useUser = () => {
  const authenticateUser = useAuthenticateUser();
  const userRbacPermissions = useUserRbacPermissions();

  return useQuery('user', async () => {
    const userStatus = await authenticateUser;
    const rawRbacPermissions = await userRbacPermissions;
    const rbacPermissions = rawRbacPermissions.map((rawPermission) => rawPermission.permission);
    const user: User = {
      accountNumber: userStatus.identity.account_number,
      canReadProducts:
        rbacPermissions.includes('subscriptions:products:read') ||
        rbacPermissions.includes('subscriptions:*:*'),
      isOrgAdmin: userStatus.identity.user.is_org_admin === true
    };
    return user;
  });
};

export { useUser as default, User };
