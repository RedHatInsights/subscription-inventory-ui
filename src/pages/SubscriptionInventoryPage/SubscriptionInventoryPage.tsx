import React, { FunctionComponent } from 'react';
import { Redirect, withRouter } from 'react-router-dom';
import Main from '@redhat-cloud-services/frontend-components/Main';
import PageHeader from '@redhat-cloud-services/frontend-components/PageHeader';
import Unavailable from '@redhat-cloud-services/frontend-components/Unavailable';
import { PageSection, Title } from '@patternfly/react-core';
import { PageHeaderTitle } from '@redhat-cloud-services/frontend-components/PageHeader';
import { useQueryClient } from 'react-query';
import { User } from '../../hooks/useUser';
import ProductsTable from '../../components/ProductsTable';
import { Processing } from '../../components/emptyState';
import useProducts from '../../hooks/useProducts';

const SubscriptionInventoryPage: FunctionComponent = () => {
  const queryClient = useQueryClient();
  const user: User = queryClient.getQueryData('user');
  const { isFetching, isLoading, error, data } = useProducts();

  const Page: FunctionComponent = () => {
    return (
      <>
        <PageHeader>
          <PageHeaderTitle title="Subscription Inventory" />
        </PageHeader>
        <Main>
          <PageSection variant="light">
            <Title headingLevel="h2">All subscriptions for account {user.accountNumber}</Title>
            <>
              {isLoading && !error && <Processing />}

              {!isLoading && !error && <ProductsTable data={data} isFetching={isFetching} />}

              {error && <Unavailable />}
            </>
          </PageSection>
        </Main>
      </>
    );
  };

  if (user.canReadProducts) {
    return <Page />;
  } else {
    return <Redirect to="/no-permissions" />;
  }
};

export default withRouter(SubscriptionInventoryPage);
