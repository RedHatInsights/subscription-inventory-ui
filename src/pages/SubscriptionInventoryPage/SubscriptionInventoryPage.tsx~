import React, { FunctionComponent } from 'react';
import { Redirect, withRouter } from 'react-router-dom';
import Main from '@redhat-cloud-services/frontend-components/Main';
import PageHeader from '@redhat-cloud-services/frontend-components/PageHeader';
import Unavailable from '@redhat-cloud-services/frontend-components/Unavailable';
import { PageSection, Split, SplitItem, Title } from '@patternfly/react-core';
import { PageHeaderTitle } from '@redhat-cloud-services/frontend-components/PageHeader';
import { useQueryClient } from 'react-query';
import { User } from '../../hooks/useUser';
import ProductsTable from '../../components/ProductsTable';
import { Processing } from '../../components/emptyState';
import useProducts from '../../hooks/useProducts';

const SubscriptionInventoryPage: FunctionComponent = () => {
  const queryClient = useQueryClient();
  const user: User = queryClient.getQueryData('user');
  const { isFetching, isLoading, error, data } = useProducts();

  const Page: FunctionComponent = () => {
    return (
      <>
        <PageHeader>
          <Split>
            <SplitItem>
              <PageHeaderTitle title="Subscription Inventory" />
            </SplitItem>
            <SplitItem isFilled></SplitItem>
            <SplitItem>Purchase subscriptions</SplitItem>
          </Split>
        </PageHeader>
        <Main>
          <PageSection variant="light">
            <Title headingLevel="h2">All subscriptions for account {user.accountNumber}</Title>
            <>
              {isLoading && !error && <Processing />}

              {!isLoading && !error && <ProductsTable data={data} isFetching={isFetching} />}

              {error && <Unavailable />}
            </>
          </PageSection>
        </Main>
      </>
    );
  };

  if (user.canReadProducts) {
    return <Page />;
  } else {
    return <Redirect to="/no-permissions" />;
  }
};

export default withRouter(SubscriptionInventoryPage);
