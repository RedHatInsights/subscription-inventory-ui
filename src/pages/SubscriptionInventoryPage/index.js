import SubscriptionInventoryPage from './SubscriptionInventoryPage';

export default SubscriptionInventoryPage;
