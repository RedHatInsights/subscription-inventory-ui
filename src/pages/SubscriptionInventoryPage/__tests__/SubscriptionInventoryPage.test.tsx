import React from 'react';
import { render, waitFor } from '@testing-library/react';
import SubscriptionInventoryPage from '../SubscriptionInventoryPage';
import Authentication from '../../../components/Authentication';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { init } from '../../../store';
import { QueryClient, QueryClientProvider } from 'react-query';
import useUser from '../../../hooks/useUser';
import useProducts from '../../../hooks/useProducts';
import { get, def } from 'bdd-lazy-var';

jest.mock('../../../hooks/useUser');
jest.mock('../../../hooks/useProducts');
jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as Record<string, unknown>),
  useLocation: () => ({
    pathname: '/'
  })
}));

const queryClient = new QueryClient();

const PageContainer = () => (
  <QueryClientProvider client={queryClient}>
    <Authentication>
      <Provider store={init().getStore()}>
        <Router>
          <SubscriptionInventoryPage />
        </Router>
      </Provider>
    </Authentication>
  </QueryClientProvider>
);

const mockAuthenticateUser = (
  orgAdminStatus: boolean,
  isError: boolean,
  canReadProducts: boolean
) => {
  const user = {
    accountNumber: '8675309',
    canReadProducts: canReadProducts,
    isOrgAdmin: orgAdminStatus,
    isSCACapable: true
  };
  (useUser as jest.Mock).mockReturnValue({
    isLoading: false,
    isFetching: false,
    isSuccess: true,
    isError: isError,
    data: user
  });

  if (isError === false) {
    queryClient.setQueryData('user', user);
  }
};

jest.mock('../../../components/ProductsTable', () => () => <div>Products Table</div>);
jest.mock('../../NoPermissionsPage', () => () => <div>Not Authorized</div>);
jest.mock('../../../components/PurchaseModal/onlineIcon.svg', () => 'Online Icon');
jest.mock('../../../components/PurchaseModal/salesIcon.svg', () => 'Sales Icon');
jest.mock('../../../components/PurchaseModal/partnersIcon.svg', () => 'Partners Icon');
jest.mock('../../../components/PurchaseModal/trainingIcon.svg', () => 'Training Icon');
jest.mock('../../../components/PurchaseModal/tryIcon.svg', () => 'Try Icon');

describe('SubscriptionInventoryPage', () => {
  def('orgAdmin', () => false);
  def('userError', () => false);
  def('canReadProducts', () => true);
  def('productsLoading', () => false);
  def('productsError', () => false);

  beforeEach(() => {
    window.insights = {};
    jest.resetAllMocks();
    mockAuthenticateUser(get('orgAdmin'), get('userError'), get('canReadProducts'));
    (useProducts as jest.Mock).mockReturnValue({
      isLoading: get('productsLoading'),
      error: get('productsError'),
      data: []
    });
  });

  it('renders correctly', async () => {
    const { container } = render(<PageContainer />);
    await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });

  describe('when the user call fails', () => {
    def('userError', () => true);

    it('renders an error message when user call fails', async () => {
      const { container } = render(<PageContainer />);
      await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the user does not have proper permissions', () => {
    def('canReadProducts', () => false);

    it('redirects to not authorized page', async () => {
      const { container } = render(<PageContainer />);
      await waitFor(() => expect(useUser).toHaveBeenCalledTimes(1));
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the products call fails', () => {
    def('productsError', () => true);

    it('renders the error page', async () => {
      const { container } = render(<PageContainer />);
      await waitFor(() => expect(useProducts).toHaveBeenCalledTimes(1));
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the products are loading', () => {
    def('productsLoading', () => true);

    it('renders the loading component', async () => {
      const { container } = render(<PageContainer />);
      await waitFor(() => expect(useProducts).toHaveBeenCalledTimes(1));
      expect(container).toMatchSnapshot();
    });
  });
});
