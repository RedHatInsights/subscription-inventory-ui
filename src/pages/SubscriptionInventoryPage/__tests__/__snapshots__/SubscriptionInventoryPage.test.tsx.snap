// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionInventoryPage renders correctly 1`] = `
<div>
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <h1
      class="pf-c-title pf-m-2xl"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe="true"
      widget-type="InsightsPageHeaderTitle"
    >
       
      Subscription Inventory
       
    </h1>
  </section>
  <section
    class="pf-l-page__main-section pf-c-page__main-section"
    page-type=""
  >
    <section
      class="pf-c-page__main-section pf-m-light"
    >
      <h2
        class="pf-c-title pf-m-xl"
        data-ouia-component-id="OUIA-Generated-Title-2"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        All subscriptions for account 
        8675309
      </h2>
      <div>
        Products Table
      </div>
    </section>
  </section>
</div>
`;

exports[`SubscriptionInventoryPage when the products are loading renders the loading component 1`] = `
<div>
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <h1
      class="pf-c-title pf-m-2xl"
      data-ouia-component-id="OUIA-Generated-Title-4"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe="true"
      widget-type="InsightsPageHeaderTitle"
    >
       
      Subscription Inventory
       
    </h1>
  </section>
  <section
    class="pf-l-page__main-section pf-c-page__main-section"
    page-type=""
  >
    <section
      class="pf-c-page__main-section pf-m-light"
    >
      <h2
        class="pf-c-title pf-m-xl"
        data-ouia-component-id="OUIA-Generated-Title-5"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        All subscriptions for account 
        8675309
      </h2>
      <div
        class="pf-l-bullseye"
      >
        <span
          aria-valuetext="Loading..."
          class="pf-c-spinner pf-m-xl"
          role="progressbar"
        >
          <span
            class="pf-c-spinner__clipper"
          />
          <span
            class="pf-c-spinner__lead-ball"
          />
          <span
            class="pf-c-spinner__tail-ball"
          />
        </span>
      </div>
    </section>
  </section>
</div>
`;

exports[`SubscriptionInventoryPage when the products call fails renders the error page 1`] = `
<div>
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <h1
      class="pf-c-title pf-m-2xl"
      data-ouia-component-id="OUIA-Generated-Title-1"
      data-ouia-component-type="PF4/Title"
      data-ouia-safe="true"
      widget-type="InsightsPageHeaderTitle"
    >
       
      Subscription Inventory
       
    </h1>
  </section>
  <section
    class="pf-l-page__main-section pf-c-page__main-section"
    page-type=""
  >
    <section
      class="pf-c-page__main-section pf-m-light"
    >
      <h2
        class="pf-c-title pf-m-xl"
        data-ouia-component-id="OUIA-Generated-Title-2"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        All subscriptions for account 
        8675309
      </h2>
      <div
        class="pf-c-empty-state pf-m-lg ins-c-empty-state__unavailable pf-m-redhat-font"
      >
        <div
          class="pf-c-empty-state__content"
        >
          <svg
            aria-hidden="true"
            class="pf-c-empty-state__icon"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
            />
          </svg>
          <h5
            class="pf-c-title pf-m-lg"
            data-ouia-component-id="OUIA-Generated-Title-3"
            data-ouia-component-type="PF4/Title"
            data-ouia-safe="true"
          >
            This page is temporarily unavailable
          </h5>
          <div
            class="pf-c-empty-state__body"
          >
            Try refreshing the page. If the problem persists, contact your organization administrator or visit our
            <a
              href="https://status.redhat.com/"
              rel="noopener noreferrer"
              target="_blank"
            >
               status page
            </a>
             for known outages.
          </div>
        </div>
      </div>
    </section>
  </section>
</div>
`;

exports[`SubscriptionInventoryPage when the user call fails renders an error message when user call fails 1`] = `
<div>
  <div
    class="pf-c-empty-state pf-m-lg ins-c-empty-state__unavailable pf-m-redhat-font"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <svg
        aria-hidden="true"
        class="pf-c-empty-state__icon"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        />
      </svg>
      <h5
        class="pf-c-title pf-m-lg"
        data-ouia-component-id="OUIA-Generated-Title-3"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        This page is temporarily unavailable
      </h5>
      <div
        class="pf-c-empty-state__body"
      >
        Try refreshing the page. If the problem persists, contact your organization administrator or visit our
        <a
          href="https://status.redhat.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
           status page
        </a>
         for known outages.
      </div>
    </div>
  </div>
</div>
`;

exports[`SubscriptionInventoryPage when the user does not have proper permissions redirects to not authorized page 1`] = `<div />`;
